// vite-plugin-ssr/dist/cjs/node/plugin/shared/isViteCliCall.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getViteBuildCliConfig = exports.isViteCliCall = void 0;
const utils_1 = require("../utils");
const cac_1 = require("cac");
function isViteCliCall() {
    let execPath = process.argv[1];
    (0, utils_1.assert)(execPath);
    execPath = (0, utils_1.toPosixPath)(execPath);
    return (
    // pnpm
    execPath.endsWith('/bin/vite.js') ||
        // npm & yarn
        execPath.endsWith('/.bin/vite') ||
        // Global install
        execPath.endsWith('/bin/vite'));
}
exports.isViteCliCall = isViteCliCall;
function getViteBuildCliConfig() {
    const cli = (0, cac_1.cac)('vite-plugin-ssr:vite-simulation');
    const desc = 'FAKE_CLI';
    cli
        .option('-c, --config <file>', desc)
        .option('--base <path>', desc)
        .option('-l, --logLevel <level>', desc)
        .option('--clearScreen', desc)
        .option('-d, --debug [feat]', desc)
        .option('-f, --filter <filter>', desc)
        .option('-m, --mode <mode>', desc);
    // build
    cli
        .command('build [root]', desc)
        .option('--target <target>', desc)
        .option('--outDir <dir>', desc)
        .option('--assetsDir <dir>', desc)
        .option('--assetsInlineLimit <number>', desc)
        .option('--ssr [entry]', desc)
        .option('--sourcemap', desc)
        .option('--minify [minifier]', desc)
        .option('--manifest [name]', desc)
        .option('--ssrManifest [name]', desc)
        .option('--force', desc)
        .option('--emptyOutDir', desc)
        .option('-w, --watch', desc)
        .action((root, options) => {
        const buildOptions = cleanOptions(options);
        config = {
            root,
            base: options.base,
            mode: options.mode,
            configFile: options.config,
            logLevel: options.logLevel,
            clearScreen: options.clearScreen,
            optimizeDeps: { force: options.force },
            build: buildOptions
        };
    });
    let config;
    cli.parse();
    (0, utils_1.assert)(config);
    return config;
    function cleanOptions(options) {
        const ret = { ...options };
        delete ret['--'];
        delete ret.c;
        delete ret.config;
        delete ret.base;
        delete ret.l;
        delete ret.logLevel;
        delete ret.clearScreen;
        delete ret.d;
        delete ret.debug;
        delete ret.f;
        delete ret.filter;
        delete ret.m;
        delete ret.mode;
        return ret;
    }
}
exports.getViteBuildCliConfig = getViteBuildCliConfig;
