// vite-plugin-ssr/dist/cjs/utils/assert.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addOnBeforeLogHook = exports.logPrefix = exports.getProjectError = exports.assertInfo = exports.assertWarning = exports.assertUsage = exports.assert = void 0;
const createErrorWithCleanStackTrace_1 = require("./createErrorWithCleanStackTrace");
const getGlobalObject_1 = require("./getGlobalObject");
const projectInfo_1 = require("./projectInfo");
const globalObject = (0, getGlobalObject_1.getGlobalObject)('assert.ts', {
    alreadyLogged: new Set()
});
const logPrefix = `[${projectInfo_1.projectInfo.npmPackageName}@${projectInfo_1.projectInfo.projectVersion}]`;
exports.logPrefix = logPrefix;
const internalErrorPrefix = `${logPrefix}[Bug]`;
const usageErrorPrefix = `${logPrefix}[Wrong Usage]`;
const warningPrefix = `${logPrefix}[Warning]`;
const infoPrefix = `${logPrefix}[Info]`;
const numberOfStackTraceLinesToRemove = 2;
function assert(condition, debugInfo) {
    if (condition) {
        return;
    }
    const debugStr = (() => {
        if (!debugInfo) {
            return null;
        }
        const debugInfoSerialized = typeof debugInfo === 'string' ? debugInfo : '`' + JSON.stringify(debugInfo) + '`';
        return `Debug info (this is for the ${projectInfo_1.projectInfo.projectName} maintainers; you can ignore this): ${debugInfoSerialized}`;
    })();
    const internalError = (0, createErrorWithCleanStackTrace_1.createErrorWithCleanStackTrace)([
        `${internalErrorPrefix} You stumbled upon a bug in ${projectInfo_1.projectInfo.projectName}'s source code.`,
        `Go to ${projectInfo_1.projectInfo.githubRepository}/issues/new and copy-paste this error. (The error's stack trace is usually enough to fix the problem).`,
        'A maintainer will fix the bug (usually under 24 hours).',
        `Don't hesitate to reach out as it makes ${projectInfo_1.projectInfo.projectName} more robust.`,
        debugStr
    ]
        .filter(Boolean)
        .join(' '), numberOfStackTraceLinesToRemove);
    globalObject.onBeforeLog?.();
    throw internalError;
}
exports.assert = assert;
function assertUsage(condition, errorMessage) {
    if (condition) {
        return;
    }
    const whiteSpace = errorMessage.startsWith('[') ? '' : ' ';
    const errMsg = `${usageErrorPrefix}${whiteSpace}${errorMessage}`;
    const usageError = (0, createErrorWithCleanStackTrace_1.createErrorWithCleanStackTrace)(errMsg, numberOfStackTraceLinesToRemove);
    globalObject.onBeforeLog?.();
    throw usageError;
}
exports.assertUsage = assertUsage;
function getProjectError(errorMessage) {
    const sep = errorMessage.startsWith('[') ? '' : ' ';
    const pluginError = (0, createErrorWithCleanStackTrace_1.createErrorWithCleanStackTrace)(`${logPrefix}${sep}${errorMessage}`, numberOfStackTraceLinesToRemove);
    return pluginError;
}
exports.getProjectError = getProjectError;
function assertWarning(condition, errorMessage, { onlyOnce = true, showStackTrace = false } = {}) {
    if (condition) {
        return;
    }
    const msg = `${warningPrefix} ${errorMessage}`;
    if (onlyOnce) {
        const { alreadyLogged } = globalObject;
        const key = onlyOnce === true ? msg : onlyOnce;
        if (alreadyLogged.has(key)) {
            return;
        }
        else {
            alreadyLogged.add(key);
        }
    }
    globalObject.onBeforeLog?.();
    if (showStackTrace) {
        console.warn(new Error(msg));
    }
    else {
        console.warn(msg);
    }
}
exports.assertWarning = assertWarning;
function assertInfo(condition, errorMessage, { onlyOnce }) {
    if (condition) {
        return;
    }
    const msg = `${infoPrefix} ${errorMessage}`;
    if (onlyOnce) {
        const { alreadyLogged } = globalObject;
        const key = msg;
        if (alreadyLogged.has(key)) {
            return;
        }
        else {
            alreadyLogged.add(key);
        }
    }
    globalObject.onBeforeLog?.();
    console.log(msg);
}
exports.assertInfo = assertInfo;
function addOnBeforeLogHook(onBeforeLog) {
    globalObject.onBeforeLog = onBeforeLog;
}
exports.addOnBeforeLogHook = addOnBeforeLogHook;
