// vite-plugin-ssr/dist/cjs/node/runtime/index-common.js
"use strict";
/* Track down log origins
import '../../utils/trackLogs'
//*/
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RenderErrorPage = exports.createPageRenderer = exports._injectAssets = exports.stampPipe = exports.pipeStream = exports.pipeNodeStream = exports.pipeWebStream = exports.dangerouslySkipEscape = exports.escapeInject = exports.renderPage = void 0;
var renderPage_1 = require("./renderPage");
Object.defineProperty(exports, "renderPage", { enumerable: true, get: function () { return renderPage_1.renderPage; } });
var renderHtml_1 = require("./html/renderHtml");
Object.defineProperty(exports, "escapeInject", { enumerable: true, get: function () { return renderHtml_1.escapeInject; } });
Object.defineProperty(exports, "dangerouslySkipEscape", { enumerable: true, get: function () { return renderHtml_1.dangerouslySkipEscape; } });
var stream_1 = require("./html/stream");
Object.defineProperty(exports, "pipeWebStream", { enumerable: true, get: function () { return stream_1.pipeWebStream; } });
Object.defineProperty(exports, "pipeNodeStream", { enumerable: true, get: function () { return stream_1.pipeNodeStream; } });
Object.defineProperty(exports, "pipeStream", { enumerable: true, get: function () { return stream_1.pipeStream; } });
Object.defineProperty(exports, "stampPipe", { enumerable: true, get: function () { return stream_1.stampPipe; } });
// TODO/v1-release: remove
var injectAssets__public_1 = require("./html/injectAssets/injectAssets__public");
Object.defineProperty(exports, "_injectAssets", { enumerable: true, get: function () { return injectAssets__public_1.injectAssets__public; } });
// TODO/v1-release: remove
var createPageRenderer_1 = require("../createPageRenderer");
Object.defineProperty(exports, "createPageRenderer", { enumerable: true, get: function () { return createPageRenderer_1.createPageRenderer; } });
// TODO/v1-release: remove
const utils_1 = require("./utils");
const RenderErrorPage_1 = require("../../shared/route/RenderErrorPage");
const picocolors_1 = __importDefault(require("@brillout/picocolors"));
const RenderErrorPage = (...args) => {
    (0, utils_1.assertWarning)(false, [
        'This import is deprecated:',
        picocolors_1.default.red("  import { RenderErrorPage } from 'vite-plugin-ssr'"),
        'Replace it with:',
        picocolors_1.default.green("  import { RenderErrorPage } from 'vite-plugin-ssr/RenderErrorPage'")
    ].join('\n'), { onlyOnce: true, showStackTrace: true });
    return (0, RenderErrorPage_1.RenderErrorPage)(...args);
};
exports.RenderErrorPage = RenderErrorPage;
// Help Telefunc detect the user's stack
globalThis._isVitePluginSsr = true;
globalThis._isVike = true;
const utils_2 = require("./utils");
(0, utils_2.addRequireShim)();
require("./page-files/setup");
