// vite-plugin-ssr/dist/cjs/node/plugin/plugins/devConfig/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.devConfig = void 0;
const determineOptimizeDeps_1 = require("./determineOptimizeDeps");
const determineFsAllowList_1 = require("./determineFsAllowList");
const getConfigVps_1 = require("../../../shared/getConfigVps");
const addSsrMiddleware_1 = require("../../shared/addSsrMiddleware");
const utils_1 = require("../../utils");
// There doesn't seem to be a straightforward way to discriminate between `$ vite preview` and `$ vite dev`
const apply = 'serve';
const isDev = true;
function devConfig() {
    let config;
    return [
        {
            name: 'vite-plugin-ssr:devConfig',
            apply,
            async config() {
                return {
                    optimizeDeps: {
                        exclude: [
                            // We exclude the vite-plugin-ssr client to be able to use `import.meta.glob()`
                            'vite-plugin-ssr/client',
                            'vite-plugin-ssr/client/router',
                            'vite-plugin-ssr/routing',
                            // - We also exclude @brillout/json-serializer to avoid:
                            //   ```
                            //   9:28:58 AM [vite] ✨ new dependencies optimized: @brillout/json-serializer/parse
                            //   9:28:58 AM [vite] ✨ optimized dependencies changed. reloading
                            //   ```
                            // - Previously, we had to exclude @brillout/json-serializer because of pnpm, but this doesn't seem to be the case anymore
                            '@brillout/json-serializer/parse',
                            '@brillout/json-serializer/stringify'
                        ]
                    }
                };
            },
            async configResolved(config_) {
                config = config_;
                const configVps = await (0, getConfigVps_1.getConfigVps)(config);
                await (0, determineOptimizeDeps_1.determineOptimizeDeps)(config, configVps, isDev);
                await (0, determineFsAllowList_1.determineFsAllowList)(config, configVps);
            },
            configureServer() {
                (0, utils_1.markEnvAsDev)();
            }
        },
        {
            name: 'vite-plugin-ssr:devConfig:addSsrMiddleware',
            apply,
            // The SSR middleware should be last middleware
            enforce: 'post',
            configureServer: {
                order: 'post',
                handler(server) {
                    if (config.server.middlewareMode)
                        return;
                    return () => {
                        (0, addSsrMiddleware_1.addSsrMiddleware)(server.middlewares, server);
                    };
                }
            }
        }
    ];
}
exports.devConfig = devConfig;
