// vite-plugin-ssr/dist/cjs/node/runtime/renderPage/analyzePage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.analyzePage = void 0;
const determineClientEntry_1 = require("../../../shared/getPageFiles/analyzePageClientSide/determineClientEntry");
const utils_1 = require("../../../shared/page-configs/utils");
const analyzePageClientSide_1 = require("../../../shared/getPageFiles/analyzePageClientSide");
const virtualFileImportPageCode_1 = require("../../shared/virtual-files/virtualFileImportPageCode");
const analyzeClientSide_1 = require("../../../shared/getPageFiles/analyzeClientSide");
const globalContext_1 = require("../globalContext");
function analyzePage(pageFilesAll, pageConfig, pageId) {
    if (pageConfig) {
        const { isClientSideRenderable, isClientRouting } = (0, analyzeClientSide_1.analyzeClientSide)(pageConfig, pageFilesAll, pageId);
        const clientFilePath = (0, utils_1.getCodeFilePath)(pageConfig, 'client');
        const clientEntry = !isClientSideRenderable ? clientFilePath : (0, determineClientEntry_1.getVPSClientEntry)(isClientRouting);
        const clientDependencies = [];
        clientDependencies.push({
            id: (0, virtualFileImportPageCode_1.getVirtualFileIdImportPageCode)(pageConfig.pageId, true),
            onlyAssets: false,
            eagerlyImported: false
        });
        // In production we inject the import of the server virtual module with ?extractAssets inside the client virtual module
        if (!(0, globalContext_1.getGlobalContext)().isProduction) {
            clientDependencies.push({
                id: (0, virtualFileImportPageCode_1.getVirtualFileIdImportPageCode)(pageConfig.pageId, false),
                onlyAssets: true,
                eagerlyImported: false
            });
        }
        //*/
        /* TODO: remove?
        Object.values(pageConfig.configElements).forEach((configElement) => {
          if (configElement.codeFilePath) {
            const { env } = configElement
            assert(env)
            const onlyAssets = env === 'server-only'
            const eagerlyImported = env === '_routing-env'
            if (onlyAssets || eagerlyImported) {
              clientDependencies.push({
                id: configElement.codeFilePath,
                onlyAssets,
                eagerlyImported
              })
            }
          }
        })
        */
        const clientEntries = [];
        if (clientEntry) {
            clientDependencies.push({
                id: clientEntry,
                onlyAssets: false,
                eagerlyImported: false
            });
            clientEntries.push(clientEntry);
        }
        return {
            isHtmlOnly: !isClientSideRenderable,
            isClientRouting,
            clientEntries,
            clientDependencies,
            // pageFilesClientSide and pageFilesServerSide are only used for debugging
            pageFilesClientSide: [],
            pageFilesServerSide: []
        };
    }
    else {
        return (0, analyzePageClientSide_1.analyzePageClientSide)(pageFilesAll, pageId);
    }
}
exports.analyzePage = analyzePage;
