// vite-plugin-ssr/dist/cjs/node/plugin/plugins/buildConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.assertRollupInput = exports.buildConfig = void 0;
const utils_1 = require("../utils");
const virtualFileImportUserCode_1 = require("../../shared/virtual-files/virtualFileImportUserCode");
const getConfigData_1 = require("./importUserCode/v1-design/getConfigData");
const utils_2 = require("../../../shared/page-configs/utils");
const findPageFiles_1 = require("../shared/findPageFiles");
const getConfigVps_1 = require("../../shared/getConfigVps");
const virtualFileImportPageCode_1 = require("../../shared/virtual-files/virtualFileImportPageCode");
const extractAssetsQuery_1 = require("../../shared/extractAssetsQuery");
function buildConfig() {
    return {
        name: 'vite-plugin-ssr:buildConfig',
        apply: 'build',
        enforce: 'post',
        configResolved: {
            order: 'post',
            async handler(config) {
                assertRollupInput(config);
                const userInputs = normalizeRollupInput(config.build.rollupOptions.input);
                const entries = await getEntries(config);
                (0, utils_1.assert)(Object.keys(entries).length > 0);
                const input = {
                    ...entries,
                    ...userInputs
                };
                config.build.rollupOptions.input = input;
                addLogHook();
            }
        },
        config(config) {
            return {
                build: {
                    outDir: (0, utils_1.determineOutDir)(config),
                    manifest: !(0, utils_1.viteIsSSR)(config)
                }
                /* We cannot do this because of https://github.com/brillout/vite-plugin-ssr/issues/447
                plublicDir: !viteIsSSR(config),
                */
            };
        }
    };
}
exports.buildConfig = buildConfig;
async function getEntries(config) {
    const configVps = await (0, getConfigVps_1.getConfigVps)(config);
    const pageFileEntries = await getPageFileEntries(config, configVps.includeAssetsImportedByServer); // TODO/v1-release: remove
    const { pageConfigsData } = await (0, getConfigData_1.getConfigData)(config.root, false, configVps.extensions);
    (0, utils_1.assertUsage)(Object.keys(pageFileEntries).length !== 0 || pageConfigsData.length !== 0, 'At least one page should be defined, see https://vite-plugin-ssr.com/add');
    if ((0, utils_1.viteIsSSR)(config)) {
        const serverEntries = analyzeServerEntries(pageConfigsData);
        const entries = {
            pageFiles: virtualFileImportUserCode_1.virtualFileIdImportUserCodeServer,
            importBuild: resolve('dist/cjs/node/importBuild.js'),
            ...pageFileEntries,
            ...serverEntries
        };
        return entries;
    }
    else {
        let { hasClientRouting, hasServerRouting, clientEntries } = analyzeClientEntries(pageConfigsData, config);
        if (Object.entries(pageFileEntries).length > 0) {
            hasClientRouting = true;
            hasServerRouting = true;
        }
        const entries = {
            ...clientEntries,
            ...pageFileEntries
        };
        const clientRoutingEntry = resolve(`dist/esm/client/router/entry.js`);
        const serverRoutingEntry = resolve(`dist/esm/client/entry.js`);
        if (hasClientRouting) {
            entries['entries/entry-client-routing'] = clientRoutingEntry;
        }
        if (hasServerRouting) {
            entries['entries/entry-server-routing'] = serverRoutingEntry;
        }
        return entries;
    }
}
function analyzeClientEntries(pageConfigsData, config) {
    let hasClientRouting = false;
    let hasServerRouting = false;
    let clientEntries = {};
    let clientFilePaths = [];
    pageConfigsData.forEach((pageConfigData) => {
        const clientRouting = (0, utils_2.getConfigValue)(pageConfigData, 'clientRouting', 'boolean');
        if (clientRouting) {
            hasClientRouting = true;
        }
        else {
            hasServerRouting = true;
        }
        {
            const { entryName, entryTarget } = getEntryFromPageConfigData(pageConfigData, true);
            clientEntries[entryName] = entryTarget;
        }
        {
            const clientFilePath = (0, utils_2.getCodeFilePath)(pageConfigData, 'client');
            if (clientFilePath) {
                clientFilePaths.push(clientFilePath);
            }
        }
    });
    clientFilePaths = (0, utils_1.unique)(clientFilePaths);
    clientFilePaths.forEach((pageFile) => {
        const { entryName, entryTarget } = getEntryFromFilePath(pageFile, config);
        clientEntries[entryName] = entryTarget;
    });
    return { hasClientRouting, hasServerRouting, clientEntries };
}
function analyzeServerEntries(pageConfigsData) {
    const serverEntries = {};
    pageConfigsData.forEach((pageConfigData) => {
        const { entryName, entryTarget } = getEntryFromPageConfigData(pageConfigData, false);
        serverEntries[entryName] = entryTarget;
    });
    return serverEntries;
}
// Ensure Rollup creates entries for each page file, see https://github.com/brillout/vite-plugin-ssr/issues/350
// (Otherwise the page files may be missing in the client manifest.json)
async function getPageFileEntries(config, includeAssetsImportedByServer) {
    const isForClientSide = !(0, utils_1.viteIsSSR)(config);
    const fileTypes = isForClientSide ? ['.page', '.page.client'] : ['.page', '.page.server'];
    if (isForClientSide && includeAssetsImportedByServer) {
        fileTypes.push('.page.server');
    }
    let pageFiles = await (0, findPageFiles_1.findPageFiles)(config, fileTypes, false);
    const pageFileEntries = {};
    pageFiles = (0, utils_1.unique)(pageFiles);
    pageFiles.forEach((pageFile) => {
        let addExtractAssetsQuery = false;
        if (isForClientSide && pageFile.includes('.page.server.')) {
            (0, utils_1.assert)(includeAssetsImportedByServer);
            addExtractAssetsQuery = true;
        }
        const { entryName, entryTarget } = getEntryFromFilePath(pageFile, config, addExtractAssetsQuery);
        pageFileEntries[entryName] = entryTarget;
    });
    return pageFileEntries;
}
function getEntryFromFilePath(filePath, config, addExtractAssetsQuery) {
    (0, utils_1.assertPosixPath)(filePath);
    (0, utils_1.assert)(filePath.startsWith('/'));
    let entryTarget = (0, utils_1.getFilePathAbsolute)(filePath, config);
    if (addExtractAssetsQuery)
        entryTarget = (0, extractAssetsQuery_1.extractAssetsAddQuery)(entryTarget);
    let entryName = filePath;
    if (addExtractAssetsQuery)
        entryName = (0, extractAssetsQuery_1.extractAssetsAddQuery)(entryName);
    entryName = (0, utils_1.removeFileExtention)(entryName);
    entryName = prependEntriesDir(entryName);
    return { entryName, entryTarget };
}
function getEntryFromPageConfigData(pageConfigData, isForClientSide) {
    let { pageId } = pageConfigData;
    const entryTarget = (0, virtualFileImportPageCode_1.getVirtualFileIdImportPageCode)(pageId, isForClientSide);
    let entryName = pageId;
    entryName = prependEntriesDir(entryName);
    return { entryName, entryTarget };
}
function prependEntriesDir(entryName) {
    if (entryName.startsWith('/')) {
        entryName = entryName.slice(1);
    }
    (0, utils_1.assert)(!entryName.startsWith('/'));
    entryName = `entries/${entryName}`;
    return entryName;
}
function resolve(filePath) {
    (0, utils_1.assert)(filePath.startsWith('dist/'));
    // [RELATIVE_PATH_FROM_DIST] Current directory: node_modules/vite-plugin-ssr/dist/cjs/node/plugin/plugins/
    return require.resolve(`../../../../../${filePath}`);
}
function normalizeRollupInput(input) {
    if (!input) {
        return {};
    }
    // Usually `input` is an oject, but the user can set it as a `string` or `string[]`
    if (typeof input === 'string') {
        input = [input];
    }
    if (Array.isArray(input)) {
        return Object.fromEntries(input.map((input) => [input, input]));
    }
    (0, utils_1.assert)((0, utils_1.isObject)(input));
    return input;
}
function addLogHook() {
    const tty = process.stdout.isTTY && !process.env.CI; // Equals https://github.com/vitejs/vite/blob/193d55c7b9cbfec5b79ebfca276d4a721e7de14d/packages/vite/src/node/plugins/reporter.ts#L27
    if (!tty)
        return;
    let lastLog = null;
    ['stdout', 'stderr'].forEach((stdName) => {
        var methodOriginal = process[stdName].write;
        process[stdName].write = function (...args) {
            lastLog = String(args[0]);
            return methodOriginal.apply(process[stdName], args);
        };
    });
    // Exhaustive list extracted from writeLine() calls at https://github.com/vitejs/vite/blob/193d55c7b9cbfec5b79ebfca276d4a721e7de14d/packages/vite/src/node/plugins/reporter.ts
    // prettier-ignore
    const viteTransientLogs = [
        'transforming (',
        'rendering chunks (',
        'computing gzip size ('
    ];
    (0, utils_1.addOnBeforeLogHook)(() => {
        // Using viteTransientLogs is very conservative as clearing the current line is low risk. (We can assume that important messages, such as errors, include a trailing new line. Usually, only transient messages have no trailing new lines.)
        if (viteTransientLogs.some((s) => lastLog?.startsWith(s))) {
            process.stdout.clearLine(0);
            process.stdout.cursorTo(0);
        }
    });
}
function assertRollupInput(config) {
    const userInputs = normalizeRollupInput(config.build.rollupOptions.input);
    const htmlInputs = Object.values(userInputs).filter((entry) => entry.endsWith('.html') || entry.endsWith('.htm'));
    const htmlInput = htmlInputs[0];
    (0, utils_1.assertUsage)(htmlInput === undefined, `The entry ${htmlInput} of config build.rollupOptions.input is an HTML entry which is forbidden when using vite-plugin-ssr, instead follow https://vite-plugin-ssr.com/add`);
}
exports.assertRollupInput = assertRollupInput;
